From host sync docs with examples (#458)
loft-sh/vcluster-docs: vcluster/_fragments/sync-from-host-secret-example.mdx

import Highlight from "@site/src/components/Highlight/Highlight";

import Flow, { Step } from "@site/src/components/Flow";

## From host Secret sync example

This guide shows how to sync Kubernetes Secret from host clusters and how you can use them in your workload running inside virtual clusters.

> ### Prerequisites

Reviewer: Prerequisites can be moved to the main section. See the `secrets.mdx` review.

- kubectl with access to the host cluster and virtual cluster.

### Set up cluster contexts

Setting up the host and virtual cluster contexts makes it easier to switch
between them.

```bash
export HOST_CTX="your-host-context"
export VCLUSTER_CTX="vcluster-ctx"
```

> then, create `foobar` namespace in your host cluster:

Reviewer: ```suggestion
then, create a namespace in your host cluster, we use `foobar` as an example:
```

```bash
kubectl --context="${HOST_CTX}" create namespace foobar
```

:::tip
You can find your contexts by running `kubectl config get-contexts`
:::

### Enable from host syncing for Secret

Enable the from host syncing for secrets in your virtual cluster configuration:

```yaml title="Enable from host syncing for a Secret"
sync:
  fromHost:
    secrets:
      enabled: true
      selector:
        mappings:
          "foobar/shared-user-env": "my-namespace/user-env"
```

This configuration:

- Enables from host syncing of the Secret `shared-user-env` in the namespace `foobar`.
- Automatically configures RBAC permissions for vCluster, so it can access this Secret (you need to re-deploy vCluster for it to take place)
- Makes this Secret accessible as `user-env` in the `my-namespace` in your vCluster.

:::tip create virtual cluster
Create or update a `virtual Cluster` following the [vCluster quick start
guide](/vcluster/#deploy-vcluster).
:::

> ### Sync Secret to virtual and use it in pod

Reviewer: ```suggestion
### Sync Secret to virtual cluster and use it in pod
```

<Flow id="secrets-from-host-example">
  <Step>
    First, create a Secret which you want to sync in the host cluster:

    ```bash title="Create Secret in the host"
    kubectl --context="${HOST_CTX}" create secret generic shared-user-env \
    --namespace=foobar \
    --from-literal=MY_ENV_1=foo \
    --from-literal=MY_ENV_2=bar
    ```
  </Step>

>   #### Ensure that Secret got synced to the vCluster

Reviewer: ```suggestion
  #### Ensure that Secret got synced to the virtual cluster
```

  Your Secret should be now accessible in the virtual cluster.
  Keep in mind, that any edit made in the virtual object is overwritten by the host object data.

  <Step>
    Check if the Secret is accessible in the virtual cluster:

    ```bash title="Get synced Secret"
    kubectl --context="${VCLUSTER_CTX}" get secrets --namespace my-namespace user-env -o yaml
    ```

    you should see similar output:

    ```yaml title="Secret contents"
    apiVersion: v1
    data:
      MY_ENV_1: Zm9v
      MY_ENV_2: YmFy
    kind: Secret
    metadata:
      creationTimestamp: "2025-02-17T12:12:59Z"
      name: user-env
      namespace: my-namespace
      resourceVersion: "15919"
      uid: 26cc03d6-5cde-4090-9f9d-3e6fcbeefd24
    type: Opaque
    ```

  </Step>

>   #### Use it in your workload

Reviewer: Steps titles should not be markdown headers.

  <Step>

    Now, you can create a new pod in the `my-namespace` namespace and specify this Secret as a source for environment variables.

    Save this pod locally to the file called `pod.yaml`:
    ```yaml title="pod.yaml"
    apiVersion: v1
    kind: Pod
    metadata:
      name: my-pod
      namespace: my-namespace
    spec:
      containers:
      - name: busybox
        image: busybox
        command:
        - sleep
        - "inf"
        env:
        # Plain Text ENV
        - name: DEMO_GREETING
          value: "Hello from the environment"
          envFrom:
          - secretRef:
            name: user-env
    ```

    then, create it in the virtual cluster:

    ```bash title="Create pod"
    kubectl --context="${VCLUSTER_CTX}" create -f pod.yaml
    ```

  </Step>
  <Step>

>     #### Verify pod environment variables

Reviewer: Steps titles should not be markdown headers.

    <Highlight color="green">Virtual Cluster</Highlight> Wait for pod running

    ```bash title="Wait for pod running"
    kubectl --context="${VCLUSTER_CTX}" wait --for=condition=ready pod my-pod --namespace my-namespace --timeout=300s
    ```

    once it is running, you can list pod's environment variables:

    ```bash title="Check environment variables"
    kubectl --context="${VCLUSTER_CTX}" exec -it --namespace my-namespace my-pod -- printenv | grep "MY_ENV"
    ```
    you should see environment successfully injected from the Secret:

    ```bash title="Environment variables from Secret"
    MY_ENV_1=foo
    MY_ENV_2=bar
    ```

  </Step>

  <Step>

>     #### Summary

Reviewer: Steps titles should not be markdown headers.
